src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py

```diff
@@ -718,16 +718,28 @@ def retrieveBlocks(self):
                 updateBlocksDAO.execute(blocks = loadedBlocks,
                                         conn = myThread.transaction.conn,
                                         transaction = True)
-            except WMException:
-                myThread.transaction.rollback()
-                raise
             except Exception as ex:
                 myThread.transaction.rollback()
-                msg =  "Unhandled exception while finished closed blocks in DBSBuffer\n"
-                msg += str(ex)
-                logging.error(msg)
-                logging.debug("Blocks for Update: %s\n" % loadedBlocks)
-                raise DBSUploadException(msg)
+                # possible deadlock with PhEDExInjector, retry once after 10s
+                time.sleep(10)
+                try:
+                    myThread.transaction.begin()
+                    updateFilesDAO.execute(blocks = loadedBlocks, status = "InDBS",
+                                           conn = myThread.transaction.conn,
+                                           transaction = True)
+                    updateBlocksDAO.execute(blocks = loadedBlocks,
+                                            conn = myThread.transaction.conn,
+                                            transaction = True)
+                except Exception as ex:
+                    myThread.transaction.rollback()
+                    msg =  "Unhandled exception while finished closed blocks in DBSBuffer\n"
+                    msg += str(ex)
+                    logging.error(msg)
+                    logging.debug("Blocks for Update: %s\n", loadedBlocks)
+                    raise DBSUploadException(msg)
+                else:
+                    myThread.transaction.commit()
```

Comment:
Arguable :-) I'm all in favor of avoiding duplicate code as much as possible (even 2 lines only).